### PR TITLE
Respect all the indexes from all the schemas in table

### PR DESCRIPTION
**Pull request by famence**
### Summary:
Support single table GSIs defined in different models/schemas within single table by merge all models' and schemas' attributes in `getCreateTableAttributeParams`

#### Table
Create table with multiple models:
```
const pk = {"type": String, "hashKey": true};
const Cat = dynamoose.model("Cat", {"pk": pk, "pk1": {"type": String, "index": {"name": "gsi1"}}});
const Dog = dynamoose.model("Cat", {"pk": pk, "pk2": {"type": String, "index": {"name": "gsi2"}}});

const table = new dynamoose.Table("Pets", [catModel, dogModel]);
await table.create({"return": "request"});
```
result (`aws:dynamodb:createTable:request`):
```
{
  "TableName": "Pets",
  "ProvisionedThroughput": {
    "ReadCapacityUnits": 1,
    "WriteCapacityUnits": 1
  },
  "AttributeDefinitions": [
    {
      "AttributeName": "pk",
      "AttributeType": "S"
    },
    {
      "AttributeName": "pk1",
      "AttributeType": "S"
    },
    {
      "AttributeName": "pk2",
      "AttributeType": "S"
    }
  ],
  "KeySchema": [
    {
      "AttributeName": "pk",
      "KeyType": "HASH"
    }
  ],
  "GlobalSecondaryIndexes": [
    {
      "IndexName": "gsi1",
      "KeySchema": [
        {
          "AttributeName": "pk1",
          "KeyType": "HASH"
        }
      ],
      "Projection": {
        "ProjectionType": "ALL"
      },
      "ProvisionedThroughput": {
        "ReadCapacityUnits": 1,
        "WriteCapacityUnits": 1
      }
    },
    {
      "IndexName": "gsi2",
      "KeySchema": [
        {
          "AttributeName": "pk2",
          "KeyType": "HASH"
        }
      ],
      "Projection": {
        "ProjectionType": "ALL"
      },
      "ProvisionedThroughput": {
        "ReadCapacityUnits": 1,
        "WriteCapacityUnits": 1
      }
    }
  ]
}
```
#### Model
Create model with multiple schemas:
```
const pk = {
  "type": String,
  "hashKey": true,
  "index": {"name": "primary"}
};
const schema1 = new dynamoose.Schema({
  "pk": pk,
  "pk1": {
    "type": String,
    "index": {"name": "gsi1"}
  }
});
const schema2 = new dynamoose.Schema({
  "pk": pk,
  "pk2": {
    "type": String,
    "index": {"name": "gsi2"}
  }
});
const User = dynamoose.model("User", [schema1, schema2]);
await model.table().create({"return": "request"});
```
result (`aws:dynamodb:createTable:request`):
```
{
  "TableName": "User",
  "ProvisionedThroughput": {
    "ReadCapacityUnits": 1,
    "WriteCapacityUnits": 1
  },
  "AttributeDefinitions": [
    {
      "AttributeName": "pk",
      "AttributeType": "S"
    },
    {
      "AttributeName": "pk1",
      "AttributeType": "S"
    },
    {
      "AttributeName": "pk2",
      "AttributeType": "S"
    }
  ],
  "KeySchema": [
  {
    "AttributeName": "pk",
    "KeyType": "HASH"
  }
  ],
  "GlobalSecondaryIndexes": [
    {
      "IndexName": "primary",
      "KeySchema": [
        {
        "AttributeName": "pk",
        "KeyType": "HASH"
        }
      ],
      "Projection": {
        "ProjectionType": "ALL"
      },
      "ProvisionedThroughput": {
        "ReadCapacityUnits": 1,
        "WriteCapacityUnits": 1
      }
    },
      {
        "IndexName": "gsi1",
        "KeySchema": [
          {
            "AttributeName": "pk1",
            "KeyType": "HASH"
          }
        ],
        "Projection": {
          "ProjectionType": "ALL"
        },
        "ProvisionedThroughput": {
        "ReadCapacityUnits": 1,
        "WriteCapacityUnits": 1
      }
    },
    {
      "IndexName": "gsi2",
      "KeySchema": [
        {
          "AttributeName": "pk2",
          "KeyType": "HASH"
        }
      ],
      "Projection": {
        "ProjectionType": "ALL"
      },
      "ProvisionedThroughput": {
        "ReadCapacityUnits": 1,
        "WriteCapacityUnits": 1
      }
    }
  ]
}
```

### GitHub linked issue:
closes https://github.com/dynamoose/dynamoose/issues/1508

### Type (select 1):
- [ ] Bug fix
- [x] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [ ] No
- [x] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [ ] Yes
- [x] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/main/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

### CODE REVIEW

**Comment by famence** [comment]
Im not sure is it a breaking change or not.
Also Im not sure if `Table.getHashKey` and `Table.getRangeKey` methods needs to walk all the schemas too. As i guess that every model in the table should have the same hash and range keys since it is just by design in DynamoDB. If im wrong about that i will implement it

File: packages/dynamoose/lib/utils/merge_objects.ts [hunk 1]
```diff
@@ -58,7 +61,22 @@ const main = (settings: MergeObjectsSettings = {"combineMethod": MergeObjectsCom
...
+	const sourceKeyBy = keyBy(source, keyByIteratee);
+	const merged = main({"combineMethod": MergeObjectsCombineMethod.ObjectCombine})<(typeof target)[0]>({}, targetKeyBy, sourceKeyBy);
+
+	return Object.values(merged);
+};
+
 const returnObject: any = main();
 returnObject.main = main;
+returnObject.MergeObjectsCombineMethod = MergeObjectsCombineMethod;
+returnObject.schemaAttributesMerger = schemaAttributesMerger;
```
Comment (famence) [inline packages/dynamoose/lib/utils/merge_objects.ts:80]: Maybe we should move it from here? Any ideas where to?

File: packages/dynamoose/lib/utils/keyBy.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import {AnySimpleObject, AnySimpleValue} from "../Types";
+
+// Source: https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore
```
Comment (fishcharlie) [inline packages/dynamoose/lib/utils/keyBy.ts:3]: Can't do this. Their license is MIT (https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore/blob/master/LICENSE) which conflicts with ours.

**Review by fishcharlie — CHANGES_REQUESTED** [review]

**Comment by famence** [comment]
@fishcharlie hello! Sorry, I was busy escaping from Russia :)
I pushed the rewritten key By util function, covered it with tests and removed the Source comment. Check it out

**Comment by talhabalaj** [comment]
Any idea when this PR is going to be merged?

**Pull request merged**